### PR TITLE
Issue #4245 standalone offline pretraining checkpoint manifest lane

### PR DESCRIPTION
**Pull request by ll7**
## Reviewer-critical summary

What changed: adds the issue #4245 standalone SAC offline pretraining lane (`scripts/training/pretrain_offline_policy.py`) and manifest-driven online fine-tune lane (`scripts/training/offline_to_online_finetune.py`) on top of the existing #4012 offline-online primitives. The new contract module writes and validates `offline-policy-checkpoint-manifest.v1` and `offline-to-online-finetune-manifest.v1` with checkpoint, normalizer sidecar, dataset, config, and environment-space hashes.

Why now: this closes the deferred #4170 maintainer request for durable offline checkpoint provenance before #4244 can consume a predeclared pretrained SAC arm.

Claim boundary: CPU smoke/provenance integrity only. This PR is not benchmark evidence, not a scratch-SAC comparison, and not a paper/dissertation claim.

Required validation: focused manifest/CLI tests, Ruff on changed Python files, both CPU smoke CLIs, JSON evidence parse, and PR readiness. Compact smoke evidence is under `docs/context/evidence/issue_4245_offline_pretrain_finetune_smoke/`.

Remaining blocker: #4244 still needs to wire any comparison arm after this PR is reviewed; full benchmark/Slurm/GPU execution remains outside this slice.

## Contract delta

New schema fields and contracts:
- `offline-policy-checkpoint-manifest.v1`: checkpoint path/hash, normalizer sidecar path/hash, training config hash, dataset path/manifest hashes, `RLTrajectoryDataset.v1` identity, offline step metadata, environment observation/action fingerprints, and `eligible_for_claim=false`.
- `offline-to-online-finetune-manifest.v1`: parent manifest hash, inherited dataset identity, loaded checkpoint hash, fine-tune config hash, output checkpoint/normalizer hashes, online timesteps, environment fingerprints, and `eligible_for_claim=false`.

New fail-closed blockers:
- missing checkpoint, normalizer sidecar, config, dataset, or dataset manifest file;
- checksum mismatch for any manifest-referenced file;
- missing dataset identity or dataset SHA;
- observation/action environment fingerprint mismatch before fine-tune.

Compatibility impact: additive scripts/config/tests only. Existing #4012 warm-start and scratch-comparator paths are not changed. No transient queue-routing state, target-host state, raw checkpoint bytes, raw datasets, or normalizer bytes are committed.

## Issue

Closes the implementation slice for https://github.com/ll7/robot_sf_ll7/issues/4245. Parent lane remains #4012; #4244 comparison work remains separate.

## Validation

- `uv run --extra training pytest tests/training/test_offline_online_rl.py tests/training/test_hybrid_replay_buffer.py tests/training/test_train_sac_sb3_offline_online.py tests/training/test_offline_pretraining_manifest.py tests/training/test_pretrain_offline_policy.py tests/training/test_offline_to_online_finetune.py -q` -> passed, 28 tests.
- `uv run ruff check robot_sf/training/offline_pretraining_manifest.py scripts/training/pretrain_offline_policy.py scripts/training/offline_to_online_finetune.py tests/training/test_offline_pretraining_manifest.py tests/training/test_pretrain_offline_policy.py tests/training/test_offline_to_online_finetune.py` -> passed.
- `uv run --extra training python scripts/training/pretrain_offline_policy.py --config configs/training/offline_online_rl/issue_4245_pretrain_smoke.yaml --output-dir output/issue_4245_offline_pretrain_smoke --manifest-out output/issue_4245_offline_pretrain_smoke/offline_checkpoint_manifest.json` -> passed.
- `uv run --extra training python scripts/training/offline_to_online_finetune.py --config configs/training/offline_online_rl/issue_4245_finetune_smoke.yaml --pretrained-manifest output/issue_4245_offline_pretrain_smoke/offline_checkpoint_manifest.json --output-dir output/issue_4245_offline_to_online_finetune_smoke --manifest-out output/issue_4245_offline_to_online_finetune_smoke/finetune_manifest.json` -> passed.
- `python -m json.tool docs/context/evidence/issue_4245_offline_pretrain_finetune_smoke/{pretrain_manifest_summary.json,finetune_manifest_summary.json,provenance_chain.json}` -> passed.
- `BASE_REF=origin/main scripts/dev/pr_ready_check.sh` -> passed core readiness before commit, 1766 tests; final mode requires this PR body and is rerun before opening.

Explicitly not run: no full benchmark campaign, no Slurm/GPU submission, no scratch-SAC comparison, no paper/dissertation claim edits.

## Downstream propagation

No issue comment was added because authorization has `comment_issue_or_pr: false`. The PR body is the single durable propagation surface for this slice; high-churn issue-state/comment updates are intentionally avoided.

<details>
<summary>Appendix: governance and artifact notes</summary>

- Fragmentation guard: no open PR covers #4245; open search only found unrelated #4251 for #4244. No merged #4245 PRs found in the dedupe search.
- Late guidance check: issue #4245 and comments were rechecked immediately before PR opening; latest maintainer comment remains `2026-07-03T09:18:32Z`, already incorporated.
- Evidence artifact policy: committed evidence contains manifest summaries and SHA-256 values only. Generated checkpoints, normalizer sidecars, datasets, and full manifests remain ignored under `output/`.
- Out of scope: full benchmark campaign, Slurm/GPU submission, scratch-SAC comparison, #4244 matrix inclusion, paper/dissertation claim edits.

</details>

### CODE REVIEW

**Comment by ll7** [comment]
## Gate review — findings (REVIEW → IMPROVE/FIX → MERGE) — HARDENED

Standalone offline-SAC pretraining + manifest-driven online fine-tune lane for #4245 (1373 add, 14 files) → runtime-adjacent training code, hardened checklist applied.

**Findings**

1. **[FIXED in this gate] `docs-evidence-integrity` failed on a structurally incompatible `SHA256SUMS`.** The committed `SHA256SUMS` checksummed six artifacts under the git-ignored `output/` tree (checkpoints, normalizer sidecars, generated manifests) that are intentionally *not* committed — so the integrity guard resolved every target as "missing" (it requires SHA256SUMS targets to be committed files, like the accepted `issue_852/.../SHA256SUMS`). This can never pass in CI, where `output/` never exists. Fixed on the branch (commit `9e237c3c7`): removed the redundant `SHA256SUMS` — **all six SHA-256 values are already recorded durably in `provenance_chain.json`** (`offline_checkpoint_sha256`, `finetune_checkpoint_sha256`, etc.) plus the two manifest summaries, which is the repo-consistent way to record hashes of intentionally-uncommitted output artifacts (same pattern as merged #4258/#4254). No provenance lost.

2. **[FIXED in this gate] Evidence files unregistered in `docs/context/catalog.yaml`.** Registered README, pretrain/finetune manifest summaries, `provenance_chain.json`, and `smoke_report.md` (`status/freshness: evidence`, `area: learned_policy`). Local `check_docs_evidence_integrity.py --base-ref origin/main` → **14 changed file(s) passed**.

**Hardened verification**
- Manifest module is fail-closed provenance machinery: `validate_offline_checkpoint_manifest` requires checkpoint/normalizer/config/dataset files to exist, rejects checksum mismatch, and requires dataset identity + dataset SHA; both manifests set `eligible_for_claim=false`. No unsafe deserialization introduced (no `pickle`/`eval`/`exec`; `current_git_commit` uses fixed-arg `subprocess.run` with error handling; model load is standard SB3).
- Manifest tests pass locally (5): required-files validation, checksum-mismatch rejection, missing-checkpoint-path, dataset-identity requirement. The pretrain/finetune CLI tests require the `training` (SB3) extra present in CI. All three test files use `tmp_path` with no cwd-relative or hardcoded paths.
- Claim boundary: CPU smoke / provenance-integrity only; checkpoints/datasets stay under ignored `output/`; not benchmark evidence, not a scratch-SAC comparison, not a paper claim.
- Closure discipline: body keeps #4245 open ("#4244 still needs to wire any comparison arm") — this is the provenance slice, not a benchmark result. Correct. Note: PR #4251 (merged this batch) already pre-registers `offline_online_sac` behind the `sac_after_issue_4245_standalone_offline_pretraining` gate, so this lane is the prerequisite that gate references.

Merge pending green CI on the fix commit. Will append to the human spot-check queue on merge (runtime-adjacent training).

**Comment by ll7** [comment]
## Gate review — findings (1387 lines → hardened care)

**Scope:** standalone offline-pretraining checkpoint-manifest lane (#4245): new module + two scripts + two smoke configs + registered evidence.

**Findings (all non-blocking / verified):**
1. **Standalone additive lane.** Every changed file is ADDED except `catalog.yaml` (additive). Scripts import existing `train_sac_sb3` primitives but do not modify them; #4012 warm-start/scratch paths untouched. No shared default changed.
2. **Provenance integrity — real & self-consistent (cross-checked SHAs).** finetune `parent_checkpoint_sha256` == pretrain `checkpoint_sha256` (`ca525330…`); `provenance_chain.json` links both checkpoint SHAs; obs/action env fingerprints identical across all three files; `git_commit_at_smoke` matches a real branch-ancestry commit. All `eligible_for_claim=false`. Genuine reproducible smoke outputs, not fabricated. No checkpoint/dataset bytes committed — only summaries + SHAs.
3. **Fail-closed wired.** `validate_offline_checkpoint_manifest`/`validate_finetune_manifest` raise `ValueError` on missing file, checksum mismatch, missing dataset identity, schema mismatch, missing env fingerprint; `assert_environment_compatible` hard-rejects fingerprint mismatch; tests cover each.
4. **All 5 evidence files registered** in `catalog.yaml` (`area: learned_policy`, `status: evidence`). No overclaim (`claim_boundary: diagnostic provenance integrity only; not benchmark evidence`).
5. **"Closes the implementation slice for #4245"** is a legitimate slice boundary: parent #4012 stays open, scratch-SAC comparison deferred to #4244 — matches issue AC. Not an overclaiming Closes.
6. Test path resolution clean (`tmp_path`/`monkeypatch`, no real SB3 training).

**Non-blocking nit:** `pretrain_offline_policy.py:923` smoke-dataset generator's terminal-flag heuristic (`range(max(min_transitions,2))` + force-terminal last step) is fragile if reused; does not affect the committed (validated) evidence.

No blocking findings. Tests: 8 passed (`--extra training`). Merge pending green CI.
